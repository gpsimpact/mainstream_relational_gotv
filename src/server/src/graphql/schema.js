import { makeExecutableSchema } from 'graphql-tools';
import resolvers from './resolvers';
import typeDefs from './types';

// create a schema
const schema = makeExecutableSchema({ typeDefs, resolvers });

export default schema;
