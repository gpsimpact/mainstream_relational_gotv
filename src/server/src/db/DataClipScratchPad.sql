-- ORG SPECIFIC COHORT
SELECT 
  u.first_name AS ambassador_first_name,
  u.last_name AS ambassador_last_name,
  u.email AS ambassador_email,
  pv.first_name AS contact_first_name,
  pv.last_name AS contact_last_name,
  pv.city as contact_city,
  vf.state_file_id AS voter_file_state_file_id,
  vf.first_name AS voter_file_first_name,
  vf.middle_name AS voter_file_middle_name,
  vf.last_name AS voter_file_last_name,
  vf.home_address AS voter_file_home_address,
  vf.city AS voter_file_city,
  vf.state AS voter_file_state,
  vf.zipcode AS voter_file_zipcode,
  vf.dob AS voter_file_dob,
  vo_ab_requested_primary,
  vo_ab_requested_date_primary,
  vo_voted_primary,
  vo_voted_date_primary,
  vo_voted_method_primary,
  vo_ab_requested_general,
  vo_ab_requested_date_general,
  vo_voted_general,
  vo_voted_date_general,
  vo_voted_method_general,
  propensity_score,
  party
FROM users u
LEFT JOIN permissions p ON u.email = p.email
LEFT JOIN potential_voters pv ON pv.user_email = u.email
LEFT JOIN shared.voter_file vf ON pv.state_file_id = vf.state_file_id
WHERE p.org_id = 'de869734-9f2d-4012-bef2-bbf0911335e2'
AND p.permission = 'AMBASSADOR'
AND pv.deleted = FALSE;


-- MASTER ADMIN VIEW
SELECT 
  o.name AS org_name,
  u.first_name AS ambassador_first_name,
  u.last_name AS ambassador_last_name,
  u.email AS ambassador_email,
  pv.first_name AS contact_first_name,
  pv.last_name AS contact_last_name,
  pv.city as contact_city,
  vf.state_file_id AS voter_file_state_file_id,
  vf.first_name AS voter_file_first_name,
  vf.middle_name AS voter_file_middle_name,
  vf.last_name AS voter_file_last_name,
  vf.home_address AS voter_file_home_address,
  vf.city AS voter_file_city,
  vf.state AS voter_file_state,
  vf.zipcode AS voter_file_zipcode,
  vf.dob AS voter_file_dob,
  vo_ab_requested_primary,
  vo_ab_requested_date_primary,
  vo_voted_primary,
  vo_voted_date_primary,
  vo_voted_method_primary,
  vo_ab_requested_general,
  vo_ab_requested_date_general,
  vo_voted_general,
  vo_voted_date_general,
  vo_voted_method_general,
  propensity_score,
  party
FROM users u
LEFT JOIN permissions p ON u.email = p.email
LEFT JOIN organizations o ON o.id = p.org_id
LEFT JOIN potential_voters pv ON pv.user_email = u.email
LEFT JOIN shared.voter_file vf ON pv.state_file_id = vf.state_file_id
WHERE pv.deleted = FALSE
ORDER BY 1;

-- master list user meta
SELECT
	*,
	rank() OVER (PARTITION BY org_name ORDER BY points_earned DESC) as rank_in_org
FROM (
	SELECT 
		names.first_name,
		names.last_name,
		names.email,
		names.org_name,
		names.count as contact_count,
		COALESCE(pr.earned,0) as points_earned
	FROM (
		SELECT
		  u.first_name,
		  u.last_name,
		  u.email,
		  o.name as org_name,
		  o.id as org_id,
		  count(pv.id) as count,
		  round(avg(vf.propensity_score),2) as avg_propensity_score
		FROM (
		  SELECT
		    *
		  FROM users
		) u LEFT JOIN (
		  SELECT
		    *
		  FROM permissions
		) p ON u.email = p.email
		LEFT JOIN (
		  SELECT
		    *
		  FROM organizations
		) o ON o.id = p.org_id
		LEFT JOIN (
		  SELECT
		    *
		  FROM potential_voters pv
		  WHERE pv.deleted = false
		) pv ON pv.user_email = u.email
		LEFT JOIN (
			SELECT 
				*
			FROM shared.voter_file
		) vf ON pv.state_file_id = vf.state_file_id
		GROUP BY 1,2,3,4,5
	) names
	LEFT JOIN (
		SELECT
			*
		FROM points_rollup
	) pr on pr.user_email = names.email and pr.org_id = names.org_id
) pre
ORDER BY 4,2;

