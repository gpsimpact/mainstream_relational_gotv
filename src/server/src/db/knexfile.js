module.exports = {
  development: {
    client: 'pg',
    connection: {
      user: process.env.DATA_DB_USER,
      database: 'gonano',
      host: process.env.DATA_DB_HOST,
      password: process.env.DATA_DB_PASS,
    },
  },
  test: {
    client: 'pg',
    connection: {
      user: process.env.DATA_DB_USER,
      database: 'gonano',
      host: process.env.DATA_DB_HOST,
      password: process.env.DATA_DB_PASS,
    },
  },
  production: {
    client: 'pg',
    // must use this for production as that is how heroku sets it automatically.
    connection: process.env.DATABASE_URL,
  },
};
