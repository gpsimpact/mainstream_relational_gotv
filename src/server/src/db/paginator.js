import base64 from 'base-64';

export const paginator = async (query, page, uniqueColumn) => {
  query.orderBy(uniqueColumn, 'ASC');

  // if cursor then base64 decode
  if (page.cursor) {
    const decodedCursor = base64.decode(page.cursor);
    query.where(uniqueColumn, '>=', decodedCursor);
  }

  // then apply limit + 1
  query.limit(page.limit + 1);

  // fetch items
  let items = await query;

  let encodedCursor;
  if (items.length > page.limit) {
    // we have a next page - make a cursor
    encodedCursor = base64.encode(items[page.limit][uniqueColumn]);
    items = items.slice(0, page.limit);
  }

  return {
    items,
    response_metadata: {
      next_cursor: encodedCursor,
    },
  };
};
