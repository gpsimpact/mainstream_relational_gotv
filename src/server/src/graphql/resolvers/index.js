import { mergeResolvers } from 'merge-graphql-schemas';
import OrganizationResolver from './OrganizationResolver';
import StatusResolver from './StatusResolver';
import UserResolver from './UserResolver';
import PotentialVoterResolver from './PotentialVoter';
import VoterResolver from './VoterResolver';
import ScalarJSONResolver from './ScalarJSONResolver';
import ScalarISODatesResolver from './ScalarISODatesResolver';
import TaskResolver from './TaskResolver';

const resolvers = [
  OrganizationResolver,
  StatusResolver,
  UserResolver,
  PotentialVoterResolver,
  VoterResolver,
  ScalarJSONResolver,
  ScalarISODatesResolver,
  TaskResolver,
];

export default mergeResolvers(resolvers);
