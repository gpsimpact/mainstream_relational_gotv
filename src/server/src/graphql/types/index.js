import { mergeTypes } from 'merge-graphql-schemas';
import organizationType from './OrganizationType';
import potentialVoterType from './PotentialVoterType';
import statusType from './StatusType';
import userType from './UserType';
import voterType from './VoterType';
import taskType from './TaskType';
import scalarJSONtype from './ScalarJSONType';
import ScalarISODatesType from './ScalarISODatesType';
import PaginationType from './PaginationType';

const types = [
  organizationType,
  potentialVoterType,
  statusType,
  userType,
  voterType,
  taskType,
  scalarJSONtype,
  ScalarISODatesType,
  PaginationType,
];

// NOTE: 2nd param is optional, and defaults to false
// Only use if you have defined the same type multiple times in
// different files and wish to attempt merging them together.
export default mergeTypes(types, { all: true });
