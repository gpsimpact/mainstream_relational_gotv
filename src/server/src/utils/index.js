import faker from 'faker';
import { includes, has } from 'lodash';
import { InsufficientPermissionsError } from '../errors';

export function mapTo(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), row));
    return Array.from(group.values());
  };
}

export function mapToMany(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, []]));
    rows.forEach(row => (group.get(keyFn(row)) || []).push(row));
    return Array.from(group.values());
  };
}

export function mapToValues(keys, keyFn, valueFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), valueFn(row)));
    return Array.from(group.values());
  };
}

export const generateFakeUsers = (num, seed) => {
  faker.seed(seed); // ensures consistent result
  const fakeUsers = [];
  Array(num)
    .fill()
    .map(() => {
      fakeUsers.push({
        first_name: faker.name.firstName(),
        last_name: faker.name.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
      });
    });
  return fakeUsers;
};

export const generateFakePVs = (num, seed, user_email, org_id) => {
  faker.seed(seed); // ensures consistent result
  const fakePVs = [];
  // const vote_methods = ['early-in-person', 'mail', 'election-day', 'N/A'];
  Array(num)
    .fill()
    .map(() => {
      fakePVs.push({
        id: faker.random.uuid(),
        first_name: faker.name.firstName(),
        last_name: faker.name.lastName(),
        city: faker.address.city(),
        user_email,
        org_id,
        state_file_id: faker.random.number().toString(),
      });
    });
  return fakePVs;
};

export const generateFakeTasks = (num, seed, pv_id) => {
  faker.seed(seed); // ensures consistent result
  const fakeTasks = [];
  const status = ['COMPLETE', 'INCOMPLETE', 'INPROGRESS', 'SKIPPED'];
  Array(num)
    .fill()
    .map(() => {
      fakeTasks.push({
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id,
        form_data: JSON.stringify({}),
        point_value: faker.random.number(),
        status: status[Math.floor(Math.random() * status.length)],
        sequence: faker.random.number(),
        description: faker.commerce.productName(),
      });
    });
  return fakeTasks;
};

export const generateFakeVoters = (num, seed) => {
  faker.seed(seed); // ensures consistent result
  const fakeVoters = [];
  const vote_methods = ['early-in-person', 'mail', 'election-day', 'N/A'];
  Array(num)
    .fill()
    .map(() => {
      fakeVoters.push({
        state_file_id: faker.random.number().toString(),
        first_name: faker.name.firstName(),
        middle_name: faker.name.firstName(),
        last_name: faker.name.lastName(),
        home_address: faker.address.streetAddress(),
        city: faker.address.city(),
        state: faker.address.state(),
        zipcode: faker.address.zipCode(),
        dob_iso8601: faker.date
          .between('1917-01-01', '1999-02-01')
          .toISOString()
          .substring(0, 10),
        vo_ab_requested: faker.random.boolean(),
        vo_ab_requested_iso8601: faker.date
          .recent()
          .toISOString()
          .substring(0, 10),
        vo_voted: faker.random.boolean(),
        vo_voted_iso8601: faker.date
          .recent()
          .toISOString()
          .substring(0, 10),
        vo_voted_method: vote_methods[Math.floor(Math.random() * vote_methods.length)],
      });
    });
  return fakeVoters;
};

export const hasPermission = (viewer, service, requiredPermission, throwOnFail = false) => {
  if (!has(viewer, 'permissions')) {
    return false;
  }
  if (!includes(viewer.permissions[service], requiredPermission)) {
    if (throwOnFail) {
      throw new InsufficientPermissionsError();
    }
    return false;
  }
  return true;
};
