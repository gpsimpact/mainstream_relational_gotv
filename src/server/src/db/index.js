import knex from 'knex';
import knexfile from './knexfile';

// Eventually we want to wrap Knex to do some batching and caching, but for
// now this will do since we know none of our queries need it
console.log(process.env.NODE_ENV);
console.log(knexfile[process.env.NODE_ENV || 'development']);
export default knex(knexfile[process.env.NODE_ENV || 'development']);
