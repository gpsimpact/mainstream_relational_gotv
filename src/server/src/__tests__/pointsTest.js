import { graphql } from 'graphql';
import db from '../db';
import schema from '../graphql/schema';
import faker from 'faker';
import MakeContext from '../Context';
// import { find } from 'lodash';
// import bcrypt from 'bcrypt';
// import jwt from 'jsonwebtoken';
import { generateFakePVs, generateFakeUsers, generateFakeVoters } from '../utils';
import redisDb from '../redisClient';

beforeAll(async () => await db.migrate.latest({ directory: 'src/db/migrations' }));
beforeEach(
  async () =>
    await Promise.all([
      db.raw('TRUNCATE TABLE potential_voters CASCADE'),
      db.raw('TRUNCATE TABLE tasks CASCADE'),
      db.raw('TRUNCATE TABLE users CASCADE'),
      db.raw('TRUNCATE TABLE organizations CASCADE'),
      db.raw('TRUNCATE TABLE voter_file CASCADE'),
      redisDb.flushall(),
    ])
);
afterAll(async () => await db.destroy());

describe('Tasks', () => {
  test('point total scenario 1', async () => {
    const users = generateFakeUsers(1, 771);
    const org1 = { id: faker.random.uuid(), name: faker.company.companyName() };
    const pvs = generateFakePVs(1, 771, users[0].email, org1.id);
    const voters = [
      {
        state_file_id: 9988776655,
        vo_ab_requested_primary: true,
        vo_voted_primary: true,
        vo_ab_requested_general: true,
        vo_voted_general: true,
        propensity_score: 0,
      },
    ];
    pvs[0].state_file_id = voters[0].state_file_id;
    const tasks = [
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs[0].id,
        form_data: JSON.stringify({}),
        point_value: 1,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs[0].id,
        form_data: JSON.stringify({}),
        point_value: 1,
        status: 'INCOMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
    ];
    await db('users').insert(users[0]);
    await db('organizations').insert(org1);
    await db('potential_voters').insert(pvs);
    await db('tasks').insert(tasks);
    await db('voter_file').insert(voters);
    const userPerms = {
      [org1.id]: ['AMBASSADOR'],
    };
    const query = `
      query {
          points(
            where: {
              user_email_is: "${users[0].email}"
            }
          ) {
            items {
              organization {
                id
              }
              user {
                first_name
                last_name
                email
              }
              earned
              potential
            }
          }
        }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: { email: users[0].email, permissions: userPerms } });
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
    expect(result.data.points.items.length).toBe(1);
    expect(result.data.points.items).toContainEqual({
      organization: { id: org1.id },
      user: {
        first_name: users[0].first_name,
        last_name: users[0].last_name,
        email: users[0].email,
      },
      earned: 137,
      potential: 2,
    });
  });

  test('point total scenario 2', async () => {
    const users = generateFakeUsers(1, 771);
    const org1 = { id: faker.random.uuid(), name: faker.company.companyName() };
    const pvs = generateFakePVs(1, 771, users[0].email, org1.id);
    const voters = [
      {
        state_file_id: 9988776655,
        vo_ab_requested_primary: true,
        vo_voted_primary: false,
        vo_ab_requested_general: false,
        vo_voted_general: true,
        propensity_score: 3,
      },
    ];
    pvs[0].state_file_id = voters[0].state_file_id;
    const tasks = [
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs[0].id,
        form_data: JSON.stringify({}),
        point_value: 1,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs[0].id,
        form_data: JSON.stringify({}),
        point_value: 1,
        status: 'INCOMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
    ];
    await db('users').insert(users[0]);
    await db('organizations').insert(org1);
    await db('potential_voters').insert(pvs);
    await db('tasks').insert(tasks);
    await db('voter_file').insert(voters);
    const userPerms = {
      [org1.id]: ['AMBASSADOR'],
    };
    const query = `
      query {
          points(
            where: {
              user_email_is: "${users[0].email}"
            }
          ) {
            items {
              organization {
                id
              }
              user {
                first_name
                last_name
                email
              }
              earned
              potential
            }
          }
        }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: { email: users[0].email, permissions: userPerms } });
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  test('point total scenario 3', async () => {
    const users = generateFakeUsers(1, 771);
    const org1 = { id: faker.random.uuid(), name: faker.company.companyName() };
    const pvs = generateFakePVs(1, 771, users[0].email, org1.id);
    await db('users').insert(users[0]);
    await db('organizations').insert(org1);
    await db('potential_voters').insert(pvs);
    // await db('tasks').insert(tasks);
    // await db('voter_file').insert(voters);
    const userPerms = {
      [org1.id]: ['AMBASSADOR'],
    };
    const query = `
      query {
          points(
            where: {
              user_email_is: "${users[0].email}"
            }
          ) {
            items {
              organization {
                id
              }
              user {
                first_name
                last_name
                email
              }
              earned
              potential
            }
          }
        }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: { email: users[0].email, permissions: userPerms } });
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
  });

  test('non admin can only see their own email addy', async () => {
    const users = generateFakeUsers(2, 772);
    const org1 = { id: faker.random.uuid(), name: faker.company.companyName() };
    const pvs1 = generateFakePVs(1, 772, users[0].email, org1.id);
    const pvs2 = generateFakePVs(1, 773, users[1].email, org1.id);
    const tasks = [
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs1[0].id,
        form_data: JSON.stringify({}),
        point_value: 101,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs1[0].id,
        form_data: JSON.stringify({}),
        point_value: 188,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs1[0].id,
        form_data: JSON.stringify({}),
        point_value: 44,
        status: 'INCOMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs2[0].id,
        form_data: JSON.stringify({}),
        point_value: 101,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs2[0].id,
        form_data: JSON.stringify({}),
        point_value: 188,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs2[0].id,
        form_data: JSON.stringify({}),
        point_value: 44,
        status: 'INCOMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
    ];
    await db('users').insert(users);
    await db('organizations').insert(org1);
    await db('potential_voters').insert(pvs1);
    await db('potential_voters').insert(pvs2);
    await db('tasks').insert(tasks);
    const userPerms = {
      [org1.id]: ['AMBASSADOR'],
    };
    const query = `
      query {
          points{
            items {
              organization {
                id
              }
              user {
                first_name
                last_name
                email
              }
            }
          }
        }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: { email: users[0].email, permissions: userPerms } });
    const result = await graphql(schema, query, rootValue, context);
    // console.log(JSON.stringify(result, null, '\t'));
    expect(result.data.points.items.length).toBe(2);
    expect(result.data.points.items).toContainEqual({
      organization: { id: org1.id },
      user: {
        first_name: users[0].first_name,
        last_name: users[0].last_name,
        email: users[0].email,
      },
    });
    expect(result.data.points.items).toContainEqual({
      organization: { id: org1.id },
      user: {
        first_name: users[1].first_name,
        last_name: users[1].last_name,
        email: null,
      },
    });
  });

  test('admin can see all email addys', async () => {
    const users = generateFakeUsers(2, 773);
    const org1 = { id: faker.random.uuid(), name: faker.company.companyName() };
    const pvs1 = generateFakePVs(1, 774, users[0].email, org1.id);
    const pvs2 = generateFakePVs(1, 775, users[1].email, org1.id);
    const tasks = [
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs1[0].id,
        form_data: JSON.stringify({}),
        point_value: 101,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs1[0].id,
        form_data: JSON.stringify({}),
        point_value: 188,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs1[0].id,
        form_data: JSON.stringify({}),
        point_value: 44,
        status: 'INCOMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs2[0].id,
        form_data: JSON.stringify({}),
        point_value: 101,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs2[0].id,
        form_data: JSON.stringify({}),
        point_value: 188,
        status: 'COMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
      {
        id: faker.random.uuid(),
        form_schema: JSON.stringify({}),
        pv_id: pvs2[0].id,
        form_data: JSON.stringify({}),
        point_value: 44,
        status: 'INCOMPLETE',
        sequence: 1,
        description: faker.commerce.productName(),
      },
    ];
    await db('users').insert(users);
    await db('organizations').insert(org1);
    await db('potential_voters').insert(pvs1);
    await db('potential_voters').insert(pvs2);
    await db('tasks').insert(tasks);
    const userPerms = {
      [org1.id]: ['AMBASSADOR', 'ADMIN'],
    };
    const query = `
      query {
          points{
            items {
              organization {
                id
              }
              user {
                first_name
                last_name
                email
              }
            }
          }
        }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: { email: users[0].email, permissions: userPerms } });
    const result = await graphql(schema, query, rootValue, context);
    // console.log(JSON.stringify(result, null, '\t'));
    expect(result.data.points.items.length).toBe(2);
    expect(result.data.points.items).toContainEqual({
      organization: { id: org1.id },
      user: {
        first_name: users[0].first_name,
        last_name: users[0].last_name,
        email: users[0].email,
      },
    });
    expect(result.data.points.items).toContainEqual({
      organization: { id: org1.id },
      user: {
        first_name: users[1].first_name,
        last_name: users[1].last_name,
        email: users[1].email,
      },
    });
  });
});
