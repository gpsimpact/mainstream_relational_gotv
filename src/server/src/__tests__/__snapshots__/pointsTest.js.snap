// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks point total scenario 1 1`] = `
Object {
  "data": Object {
    "points": Object {
      "items": Array [
        Object {
          "earned": 137,
          "organization": Object {
            "id": "dfbda790-67cb-47da-a05a-021d6f30409a",
          },
          "potential": 2,
          "user": Object {
            "email": "Cindy.Pacocha@gmail.com",
            "first_name": "Geo",
            "last_name": "Franecki",
          },
        },
      ],
    },
  },
}
`;

exports[`Tasks point total scenario 2 1`] = `
Object {
  "data": Object {
    "points": Object {
      "items": Array [
        Object {
          "earned": 27,
          "organization": Object {
            "id": "dfbda790-67cb-47da-a05a-021d6f30409a",
          },
          "potential": 2,
          "user": Object {
            "email": "Cindy.Pacocha@gmail.com",
            "first_name": "Geo",
            "last_name": "Franecki",
          },
        },
      ],
    },
  },
}
`;

exports[`Tasks point total scenario 3 1`] = `
Object {
  "data": Object {
    "points": Object {
      "items": Array [
        Object {
          "earned": 0,
          "organization": Object {
            "id": "dfbda790-67cb-47da-a05a-021d6f30409a",
          },
          "potential": 0,
          "user": Object {
            "email": "Cindy.Pacocha@gmail.com",
            "first_name": "Geo",
            "last_name": "Franecki",
          },
        },
      ],
    },
  },
}
`;
